src/main/java/org/traccar/model/User.java

```diff
@@ -273,4 +277,56 @@ public boolean isPasswordValid(String password) {
         return Hashing.validatePassword(password, hashedPassword, salt);
     }
 
+    private String totpKey;
+    private boolean useTotp;
+
+    public void setTotpKey(String key) {
+        if (key.isEmpty()) {
+            this.totpKey = null;
+            this.totpKey = this.getTotpKey();
+            return;
+        }
+        this.totpKey = key;
+    }
+
+    public String getTotpKey() {
+        if (this.totpKey == null) {
+            GoogleAuthenticator gAuth = new GoogleAuthenticator();
+            this.setTotpKey(gAuth.createCredentials().getKey());
+        }
+        return this.totpKey;
+    }
+
+    public void setUseTotp(boolean doUse) {
+        if ((doUse) && (!this.useTotp)) {
+            this.totpKey = null;
+            this.getTotpKey();
+        }
+        this.useTotp = doUse;
+    }
+
+    public boolean getUseTotp() {
+        if (!Context.getConfig().getBoolean("totp.enabled")) {
+            return false;
+        }
+        return this.useTotp;
+    }
+
+    @JsonIgnore
+    @QueryIgnore
+    public boolean isTotpAuthCodeValid(int code)
+    throws GoogleAuthenticatorException {
+        if (!Context.getConfig().getBoolean("totp.enabled")) {
+            return true;
+        }
+
+        GoogleAuthenticator gAuth = new GoogleAuthenticator();
```

Comment:
Please use full words for variable names as much as possible. This can be `authenticator` for example.